File: Server~/src/tools/gameobject/DuplicateGameObjectTool.ts
import { z } from 'zod';
import { BaseTool } from '../base/BaseTool.js';
import { Tool, Tags } from '../base/ToolDecorators.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

/**
 * Input schema for duplicate_gameobject tool
 * Based on Unity API: Object.Instantiate
 * https://docs.unity3d.com/ScriptReference/Object.Instantiate.html
 */
const DuplicateGameObjectToolArgsSchema = z.object({
  instanceId: z.number().int().optional().describe('Instance ID of the GameObject to duplicate'),
  instanceIds: z.array(z.number().int()).optional().describe('Array of GameObject instance IDs to duplicate'),
  objectPath: z.string().optional().describe('Hierarchy path to the GameObject to duplicate'),
  count: z.number().int().min(1).max(100).optional().default(1).describe('Number of duplicates to create. Default: 1'),
  keepParent: z.boolean().optional().default(true).describe('Whether to keep the same parent. Default: true'),
  offsetX: z.number().optional().default(1).describe('X offset between duplicates. Default: 1'),
  offsetY: z.number().optional().default(0).describe('Y offset between duplicates. Default: 0'),
  offsetZ: z.number().optional().default(0).describe('Z offset between duplicates. Default: 0')
});

/**
 * DuplicateGameObject Tool
 * Duplicates/clones GameObjects in the scene
 */
@Tool({
  name: 'duplicate_gameobject',
  description: 'Duplicate/clone GameObjects in the scene. Supports Undo/Redo.',
  category: 'gameobject',
  version: '1.0.0'
})
@Tags(['unity', 'gameobject', 'duplicate', 'clone', 'copy'])
export class DuplicateGameObjectTool extends BaseTool {
  get name() {
    return 'duplicate_gameobject';
  }

  get description() {
    return 'Duplicate/clone GameObjects in the scene. Supports Undo/Redo.';
  }

  get inputSchema() {
    return DuplicateGameObjectToolArgsSchema;
  }

  get category() {
    return 'gameobject';
  }

  protected formatSuccessResponse(result: any): CallToolResult {
    return {
      content: [{
        type: 'text',
        text: `✅ ${result.message || 'GameObjects duplicated successfully'}`
      }]
    };
  }
}

File: Editor/Tools/GameObject/DuplicateGameObjectTool.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using UnityEditor;
using McpUnity.Utils;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool to duplicate/clone GameObjects in the scene
    /// Based on Unity API: Object.Instantiate
    /// https://docs.unity3d.com/ScriptReference/Object.Instantiate.html
    /// </summary>
    public class DuplicateGameObjectTool : McpToolBase
    {
        public DuplicateGameObjectTool()
        {
            Name = "duplicate_gameobject";
            Description = "Duplicate/clone GameObjects in the scene. Supports Undo/Redo.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            try
            {
                JArray instanceIdsArray = parameters["instanceIds"] as JArray;
                int? singleInstanceId = parameters["instanceId"]?.ToObject<int>();
                string objectPath = parameters["objectPath"]?.ToObject<string>();
                int count = parameters["count"]?.ToObject<int>() ?? 1;
                bool keepParent = parameters["keepParent"]?.ToObject<bool>() ?? true;
                
                // Position offset for multiple duplicates
                float offsetX = parameters["offsetX"]?.ToObject<float>() ?? 1f;
                float offsetY = parameters["offsetY"]?.ToObject<float>() ?? 0f;
                float offsetZ = parameters["offsetZ"]?.ToObject<float>() ?? 0f;

                List<GameObject> sourcesToDuplicate = new List<GameObject>();

                // Method 1: Single instance ID
                if (singleInstanceId.HasValue)
                {
                    GameObject obj = EditorUtility.InstanceIDToObject(singleInstanceId.Value) as GameObject;
                    if (obj != null)
                    {
                        sourcesToDuplicate.Add(obj);
                    }
                }

                // Method 2: Multiple instance IDs
                if (instanceIdsArray != null && instanceIdsArray.Count > 0)
                {
                    foreach (var id in instanceIdsArray)
                    {
                        GameObject obj = EditorUtility.InstanceIDToObject(id.ToObject<int>()) as GameObject;
                        if (obj != null && !sourcesToDuplicate.Contains(obj))
                        {
                            sourcesToDuplicate.Add(obj);
                        }
                    }
                }

                // Method 3: By object path
                if (!string.IsNullOrEmpty(objectPath))
                {
                    GameObject obj = GameObject.Find(objectPath);
                    if (obj != null && !sourcesToDuplicate.Contains(obj))
                    {
                        sourcesToDuplicate.Add(obj);
                    }
                }

                if (sourcesToDuplicate.Count == 0)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "No source GameObjects found to duplicate. Provide 'instanceId', 'instanceIds', or 'objectPath'.",
                        "validation_error");
                }

                JArray duplicatedArray = new JArray();
                int totalCreated = 0;

                foreach (GameObject source in sourcesToDuplicate)
                {
                    Transform parent = keepParent ? source.transform.parent : null;

                    for (int i = 0; i < count; i++)
                    {
                        // Calculate position with offset
                        Vector3 newPosition = source.transform.position + new Vector3(
                            offsetX * (i + 1),
                            offsetY * (i + 1),
                            offsetZ * (i + 1)
                        );

                        // Instantiate the duplicate
                        GameObject duplicate = UnityEngine.Object.Instantiate(
                            source,
                            newPosition,
                            source.transform.rotation,
                            parent
                        );

                        // Set name without (Clone) suffix
                        duplicate.name = source.name + (count > 1 ? $" ({i + 1})" : " (Copy)");

                        // Register for undo
                        Undo.RegisterCreatedObjectUndo(duplicate, "Duplicate GameObject");

                        duplicatedArray.Add(new JObject
                        {
                            ["sourceName"] = source.name,
                            ["sourceInstanceId"] = source.GetInstanceID(),
                            ["newName"] = duplicate.name,
                            ["newInstanceId"] = duplicate.GetInstanceID(),
                            ["position"] = new JObject
                            {
                                ["x"] = duplicate.transform.position.x,
                                ["y"] = duplicate.transform.position.y,
                                ["z"] = duplicate.transform.position.z
                            }
                        });
                        totalCreated++;
                    }
                }

                return new JObject
                {
                    ["success"] = true,
                    ["message"] = $"Duplicated {sourcesToDuplicate.Count} source(s), created {totalCreated} new GameObject(s).",
                    ["sourceCount"] = sourcesToDuplicate.Count,
                    ["totalCreated"] = totalCreated,
                    ["duplicatedObjects"] = duplicatedArray
                };
            }
            catch (Exception ex)
            {
                McpLogger.LogError($"DuplicateGameObjectTool error: {ex.Message}");
                return McpUnitySocketHandler.CreateErrorResponse($"Failed: {ex.Message}", "execution_error");
            }
        }
    }
}
